File: cathode-jobs/src/main/java/net/simonvt/cathode/remote/sync/shows/SyncTrendingShows.java
/*
 * Copyright (C) 2013 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.remote.sync.shows;

import android.content.ContentProviderOperation;
import android.content.ContentValues;
import android.database.Cursor;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import net.simonvt.cathode.api.entity.Show;
import net.simonvt.cathode.api.entity.TrendingItem;
import net.simonvt.cathode.api.enumeration.Extended;
import net.simonvt.cathode.api.service.ShowsService;
import net.simonvt.cathode.jobqueue.JobPriority;
import net.simonvt.cathode.provider.DatabaseContract.ShowColumns;
import net.simonvt.cathode.provider.ProviderSchematic.Shows;
import net.simonvt.cathode.provider.ShowDatabaseHelper;
import net.simonvt.cathode.remote.CallJob;
import net.simonvt.schematic.Cursors;
import retrofit2.Call;

public class SyncTrendingShows extends CallJob<List<TrendingItem>> {

  private static final int LIMIT = 50;

  @Inject transient ShowsService showsService;

  @Inject transient ShowDatabaseHelper showHelper;

  @Override public String key() {
    return "SyncTrendingShows";
  }

  @Override public int getPriority() {
    return JobPriority.SUGGESTIONS;
  }

  @Override public Call<List<TrendingItem>> getCall() {
    return showsService.getTrendingShows(LIMIT, Extended.FULL);
  }

  @Override public boolean handleResponse(List<TrendingItem> shows) {
    ArrayList<ContentProviderOperation> ops = new ArrayList<>();
    Cursor c = getContentResolver().query(Shows.SHOWS_TRENDING, null, null, null, null);

    List<Long> showIds = new ArrayList<>();
    while (c.moveToNext()) {
      final long showId = Cursors.getLong(c, ShowColumns.ID);
      showIds.add(showId);
    }
    c.close();

    for (int i = 0, count = shows.size(); i < count; i++) {
      TrendingItem item = shows.get(i);
      Show show = item.getShow();
      final long showId = showHelper.partialUpdate(show);

      ContentValues values = new ContentValues();
      values.put(ShowColumns.TRENDING_INDEX, i);
      ContentProviderOperation op =
          ContentProviderOperation.newUpdate(Shows.withId(showId)).withValues(values).build();
      ops.add(op);

      showIds.remove(showId);
    }

    for (Long showId : showIds) {
      ContentProviderOperation op = ContentProviderOperation.newUpdate(Shows.withId(showId))
          .withValue(ShowColumns.TRENDING_INDEX, -1)
          .build();
      ops.add(op);
    }

    return applyBatch(ops);
  }
}

File: cathode-sync/src/main/java/net/simonvt/cathode/actions/shows/SyncTrendingShows.kt
/*
 * Copyright (C) 2013 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.actions.shows

import android.content.ContentProviderOperation
import android.content.ContentValues
import android.content.Context
import net.simonvt.cathode.actions.CallAction
import net.simonvt.cathode.api.entity.TrendingItem
import net.simonvt.cathode.api.enumeration.Extended
import net.simonvt.cathode.api.service.ShowsService
import net.simonvt.cathode.common.database.forEach
import net.simonvt.cathode.common.database.getLong
import net.simonvt.cathode.provider.DatabaseContract.ShowColumns
import net.simonvt.cathode.provider.ProviderSchematic.Shows
import net.simonvt.cathode.provider.batch
import net.simonvt.cathode.provider.helper.ShowDatabaseHelper
import net.simonvt.cathode.provider.query
import net.simonvt.cathode.settings.SuggestionsTimestamps
import retrofit2.Call
import javax.inject.Inject

class SyncTrendingShows @Inject constructor(
  private val context: Context,
  private val showHelper: ShowDatabaseHelper,
  private val showsService: ShowsService
) : CallAction<Unit, List<TrendingItem>>() {

  override fun getCall(params: Unit): Call<List<TrendingItem>> =
    showsService.getTrendingShows(LIMIT, Extended.FULL)

  override suspend fun handleResponse(params: Unit, response: List<TrendingItem>) {
    val ops = arrayListOf<ContentProviderOperation>()
    val showIds = mutableListOf<Long>()

    val localShows = context.contentResolver.query(Shows.SHOWS_TRENDING)
    localShows.forEach { cursor -> showIds.add(cursor.getLong(ShowColumns.ID)) }
    localShows.close()

    response.forEachIndexed { index, trendingItem ->
      val show = trendingItem.show!!
      val showId = showHelper.partialUpdate(show)
      showIds.remove(showId)

      val values = ContentValues()
      values.put(ShowColumns.TRENDING_INDEX, index)
      val op = ContentProviderOperation.newUpdate(Shows.withId(showId)).withValues(values).build()
      ops.add(op)
    }

    for (showId in showIds) {
      val op = ContentProviderOperation.newUpdate(Shows.withId(showId))
        .withValue(ShowColumns.TRENDING_INDEX, -1)
        .build()
      ops.add(op)
    }

    context.contentResolver.batch(ops)

    SuggestionsTimestamps.get(context)
      .edit()
      .putLong(SuggestionsTimestamps.SHOWS_TRENDING, System.currentTimeMillis())
      .apply()
  }

  companion object {

    private const val LIMIT = 50

    fun key() = "SyncTrendingShows"
  }
}
